README.md
# PPG-Log
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/ppg-log/0.1.0?logo=python&logoColor=FFD43B)](https://pypi.org/project/ppg-log/)
[![PyPI](https://img.shields.io/pypi/v/ppg-log?logo=Python&logoColor=FFD43B)](https://pypi.org/project/ppg-log/)
[![PyPI - License](https://img.shields.io/pypi/l/ppg-log?color=magenta)](https://github.com/sco1/ppg-log/blob/main/LICENSE)
[![pre-commit.ci status](https://results.pre-commit.ci/badge/github/sco1/ppg-log/main.svg)](https://results.pre-commit.ci/latest/github/sco1/ppg-log/main)
[![Code style: black](https://img.shields.io/badge/code%20style-black-black)](https://github.com/psf/black)
[![Open in Visual Studio Code](https://img.shields.io/badge/Open%20in-VSCode.dev-blue)](https://vscode.dev/github.com/sco1/ppg-log)

Helper Tools for PPG FlySight Flight Log Management

🚨 This is an alpha project. User-facing functionality is still under development 🚨

## Installation
Install from PyPi with your favorite `pip` invocation:

```bash
$ pip install ppg-log
```

You can confirm proper installation via the `ppglog` CLI:
<!-- [[[cog
import cog
from subprocess import PIPE, run
out = run(["ppglog", "--help"], stdout=PIPE, encoding="ascii")
cog.out(
    f"```bash\n$ ppglog --help\n{out.stdout.rstrip()}\n```"
)
]]] -->
```bash
$ ppglog --help
Usage: ppglog [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  batch   Batch flight log processing pipeline.
  db      Interact with a PPG Log database instance.
  single  Single flight log processing pipeline.
```
<!-- [[[end]]] -->

## Usage
### Environment Variables
The following environment variables are provided to help customize pipeline behaviors.

| Variable Name      | Description                       | Default      |
|--------------------|-----------------------------------|--------------|
| `DB_URL`           | Flight log database URL           | `'./tmp.db'` |
| `PROMPT_START_DIR` | Start path for UI file/dir prompt | `'.'`        |

### `ppglog single`
Process a single FlySight log file.
#### Input Parameters
| Parameter              | Description                                                                  | Type         | Default    |
|------------------------|------------------------------------------------------------------------------|--------------|------------|
| `--log-filepath`       | Path to FlySight log to parse.                                               | `Path\|None` | GUI Prompt |
| `--start-trim`         | Seconds to discard from the beginning of the flight log.                     | `int\|float` | `45`       |
| `--airborne-threshold` | Minimum groundspeed, as m/s, required to be considered airborne.             | `int\|float` | `2.235`    |
| `--time-threshold`     | Duration, as seconds, used to characterize flight segments.                  | `int\|float` | `15`       |
| `--midair-start`       | Consider the start of the file as the start of a flight segment.<sup>1</sup> | `bool`       | `False`    |
| `--show-plot`          | Show parsed flight log summary plot.                                         | `bool`       | `True`     |
| `--plot_save_dir`      | Path to save parsed flight log summary plot.<sup>2</sup>                     | `Path\|None` | `None`     |
| `--db-insert`          | Insert flight logs into the currently configured database.                   | `bool`       | `False`    |

1. If `True`, any specificed `--start-trim` will be discarded
2. If `None`, the summary plot will not be saved

### `ppglog batch`
Batch process a directory of FlySight log files.
#### Input Parameters
| Parameter              | Description                                                      | Type         | Default    |
|------------------------|------------------------------------------------------------------|--------------|------------|
| `--log-dir`            | Path to FlySight log directory to parse.                         | `Path\|None` | GUI Prompt |
| `--log-pattern`        | FlySight log file glob pattern.<sup>1,2</sup>                    | `str`        | `"*.CSV"`  |
| `--start-trim`         | Seconds to discard from the beginning of the flight log.         | `int\|float` | `45`       |
| `--airborne-threshold` | Minimum groundspeed, as m/s, required to be considered airborne. | `int\|float` | `2.235`    |
| `--time-threshold`     | Duration, as seconds, used to characterize flight segments.      | `int\|float` | `15`       |
| `--plot-save-dir`      | Path to save parsed flight log summary plot.<sup>3</sup>         | `Path\|None` | `None`     |
| `--db-insert`          | Insert flight logs into the currently configured database.       | `bool`       | `False`    |
| `--verbose`            | Display in-console information on the running parsing operation. | `bool`       | `True`     |

1. Case sensitivity is deferred to the host OS
2. Recursive globbing requires manual specification (e.g. `**/*.CSV`)
3. If `None`, the summary plot will not be saved

### `ppglog db`
Subcommands for interacting with the PPG Log database.
### `ppglog db set-address`
Save the db address to a local `.env` file.

**NOTE:** If a `.env` file does not exist in the current directory, one will be created for you.
#### Input Parameters
| Parameter | Description        | Type  | Default  |
|-----------|--------------------|-------|----------|
| `--value` | db address string. | `str` | Required |

## Contributing
### Development Environment
This project uses [Poetry](https://python-poetry.org/) to manage dependencies. With your fork cloned to your local machine, you can install the project and its dependencies to create a development environment using:

```bash
$ poetry install
```

A [pre-commit](https://pre-commit.com) configuration is also provided to create a pre-commit hook so linting errors aren't committed:

```bash
$ pre-commit install
```

### Testing & Coverage
A [pytest](https://docs.pytest.org/en/latest/) suite is provided, with coverage reporting from [pytest-cov](https://github.com/pytest-dev/pytest-cov). A [tox](https://github.com/tox-dev/tox/) configuration is provided to test across all supported versions of Python. Testing will be skipped for Python versions that cannot be found.

```bash
$ tox
```

Details on missing coverage, including in the test suite, is provided in the report to allow the user to generate additional tests for full coverage.


CHANGELOG.md
# Changelog
Versions follow [Semantic Versioning](https://semver.org/spec/v2.0.0.html) (`<major>`.`<minor>`.`<patch>`)

## [v0.2.0]
### Added
* #3 Add CLI for flight log processing pipelines
* #4 Add database support for processed flight log information
* #13 Add `--midair-start` processing flag to assist with midair logging start
* #14 Add support for parsing GPX flight logs from Gaggle

### Fixed
* #4, #5, #6, #7 Fix flight segment classification logic & better account for transient noise/spikes in flight data

## [v0.1.0]
Initial release 🎉


tests/utils.py
from __future__ import annotations

from pathlib import Path

import pandas as pd

from ppg_log import metrics

KEEP_COLS = ["velN", "velE", "elapsed_time", "groundspeed", "flight_mode"]


def dump_sample_json(
    flight_data: pd.DataFrame,
    out_filepath: Path,
    keep_cols: list[str] = KEEP_COLS,
    start_idx: int | None = None,
    end_idx: int | None = None,
) -> None:  # pragma: no cover
    """
    Dump sample flight data to a JSON file specified by `out_filepath` for use with testing.

    Columns to dump may be specified by `keep_cols`, which defaults to:
        * "velN"
        * "velE"
        * "elapsed_time"
        * "groundspeed"
        * "flight_mode"

    Data to dump may be selected using `start_idx` & `end_idx`, which follow Python's `slice`
    semantics.
    """
    flight_data[keep_cols].iloc[slice(start_idx, end_idx)].to_json(out_filepath)


def json_from_file(
    log_file: pd.DataFrame,
    out_filepath: Path,
    keep_cols: list[str] = KEEP_COLS,
    start_idx: int | None = None,
    end_idx: int | None = None,
) -> metrics.FlightLog:  # pragma: no cover
    """Shortcut for `dump_sample_json` but starting from a log CSV instead of a `DataFrame`."""
    flight_log = metrics.process_log(log_file)
    dump_sample_json(
        flight_log.flight_data,
        out_filepath=out_filepath,
        keep_cols=keep_cols,
        start_idx=start_idx,
        end_idx=end_idx,
    )

    return flight_log


def dump_sample_csv(
    flight_data: pd.DataFrame,
    out_filepath: Path,
    keep_cols: list[str] = KEEP_COLS,
    start_idx: int | None = None,
    end_idx: int | None = None,
) -> None:  # pragma: no cover
    """
    Dump sample flight data to a CSV file specified by `out_filepath` for use with testing.

    Columns to dump may be specified by `keep_cols`, which defaults to:
        * "velN"
        * "velE"
        * "elapsed_time"
        * "groundspeed"
        * "flight_mode"

    Data to dump may be selected using `start_idx` & `end_idx`, which follow Python's `slice`
    semantics.
    """
    flight_data[keep_cols].iloc[slice(start_idx, end_idx)].to_csv(out_filepath)


tests/checks.py
from __future__ import annotations

import datetime as dt
import typing as t

import pandas as pd
import pytest
from pytest_check import check_func

from ppg_log import metrics


@check_func  # type: ignore[misc]  # fine with this untyped decorator
def is_col(df: pd.DataFrame, col_name: str) -> None:
    assert col_name in df.columns


@check_func  # type: ignore[misc]  # fine with this untyped decorator
def is_dtype(df: pd.DataFrame, col_name: str, type_check: t.Callable) -> None:
    """
    Check that the query dataframe matches the correct type.

    `type_check` is assumed to be a callable function that returns a boolean. The intent is to use
    pandas` built-in `pandas.api.types.is_*` functions.
    """
    assert type_check(df[col_name])


@check_func  # type: ignore[misc]  # fine with this untyped decorator
def segment_isclose(
    flight: metrics.FlightSegment,
    truth: metrics.FlightSegment,
    idx_tol: int,
    duration_tol: dt.timedelta,
) -> None:
    flight_duration = flight.duration.total_seconds()
    truth_duration = truth.duration.total_seconds()
    tol = duration_tol.total_seconds()
    duration_msg = f"Duration check failed. Segment: {flight_duration}, Truth: {truth_duration}, Tolerance: {tol}"  # noqa: E501
    assert flight_duration == pytest.approx(truth_duration, abs=tol), duration_msg

    start_idx_msg = f"Start idx check failed. Segment: {flight.start_idx}, Truth: {truth.start_idx}, Tolerance: {idx_tol}"  # noqa: E501
    end_idx_msg = f"End idx check failed. Segment: {flight.end_idx}, Truth: {truth.end_idx}, Tolerance: {idx_tol}"  # noqa: E501
    assert flight.start_idx == pytest.approx(truth.start_idx, abs=idx_tol), start_idx_msg
    assert flight.end_idx == pytest.approx(truth.end_idx, abs=idx_tol), end_idx_msg


tests/test_db.py
import datetime as dt
from functools import partial

import peewee as pw
import pytest

from ppg_log import db
from ppg_log.metrics import FlightLog, FlightSegment, LogMetadata

TEST_DB = pw.SqliteDatabase(":memory:")


DUMMY_DURATION = dt.timedelta(seconds=5)
PARTIAL_SEGMENT = partial(FlightSegment, start_idx=-1, end_idx=-1)
PARTIAL_META = partial(LogMetadata, log_date="22-04-20", log_time="04-20-00")

DUMMY_FLIGHT_LOG = FlightLog(
    flight_data=None,
    metadata=PARTIAL_META(
        n_flight_segments=1,
        total_flight_time=DUMMY_DURATION,
        flight_segments=[PARTIAL_SEGMENT(duration=DUMMY_DURATION)],
    ),
)

DUMMY_BULK_LOGS = [
    DUMMY_FLIGHT_LOG,
    FlightLog(
        flight_data=None,
        metadata=PARTIAL_META(
            log_date="22-04-21",
            n_flight_segments=1,
            total_flight_time=DUMMY_DURATION,
            flight_segments=[PARTIAL_SEGMENT(duration=DUMMY_DURATION)],
        ),
    ),
]

DUMMY_FLIGHT_LOG_NO_FLIGHTS = FlightLog(
    flight_data=None,
    metadata=PARTIAL_META(log_date="22-04-21"),
)


@pytest.fixture
def session(request: pytest.FixtureRequest) -> None:
    # We only have one table so we don't need to bind references
    TEST_DB.bind([db.FlightLogEntry], bind_refs=False, bind_backrefs=False)

    TEST_DB.connect()
    db.create_db()

    def teardown() -> None:
        TEST_DB.drop_tables(db.FlightLogEntry)
        TEST_DB.close()

    request.addfinalizer(teardown)


def test_single_insert(session: None) -> None:
    db.insert_single(DUMMY_FLIGHT_LOG)

    n_rows = db.FlightLogEntry.select(pw.fn.COUNT(db.FlightLogEntry.flight_log_id)).scalar()
    assert n_rows == 1


def test_insert_log_empty_segments(session: None) -> None:
    db.insert_single(DUMMY_FLIGHT_LOG_NO_FLIGHTS)

    n_rows = db.FlightLogEntry.select(pw.fn.COUNT(db.FlightLogEntry.flight_log_id)).scalar()
    assert n_rows == 1


def test_single_insert_duplicate_aborts(session: None, capsys: pytest.CaptureFixture) -> None:
    db.insert_single(DUMMY_FLIGHT_LOG)
    db.insert_single(DUMMY_FLIGHT_LOG)

    assert "Insertion aborted" in capsys.readouterr().out

    n_rows = db.FlightLogEntry.select(pw.fn.COUNT(db.FlightLogEntry.flight_log_id)).scalar()
    assert n_rows == 1


def test_bulk_insert(session: None) -> None:
    db.bulk_insert(DUMMY_BULK_LOGS)

    n_rows = db.FlightLogEntry.select(pw.fn.COUNT(db.FlightLogEntry.flight_log_id)).scalar()
    assert n_rows == 2


def test_bulk_insert_empty_flight(session: None) -> None:
    db.bulk_insert([DUMMY_FLIGHT_LOG, DUMMY_FLIGHT_LOG_NO_FLIGHTS])

    n_rows = db.FlightLogEntry.select(pw.fn.COUNT(db.FlightLogEntry.flight_log_id)).scalar()
    assert n_rows == 2


def test_bulk_insert_ignore_duplicates(session: None) -> None:
    db.insert_single(DUMMY_FLIGHT_LOG)
    db.bulk_insert(DUMMY_BULK_LOGS, verbose=False)

    n_rows = db.FlightLogEntry.select(pw.fn.COUNT(db.FlightLogEntry.flight_log_id)).scalar()
    assert n_rows == 2


def test_bulk_insert_ignore_duplicates_empty_db(session: None) -> None:
    db.bulk_insert([DUMMY_FLIGHT_LOG, DUMMY_FLIGHT_LOG], verbose=False)

    n_rows = db.FlightLogEntry.select(pw.fn.COUNT(db.FlightLogEntry.flight_log_id)).scalar()
    assert n_rows == 1


def test_flight_log_empty_segments() -> None:
    entry = db.FlightLogEntry.from_flight_log(DUMMY_FLIGHT_LOG_NO_FLIGHTS)

    assert entry.flight_segment_durations == ""


def test_summary_empty_db(session: None) -> None:
    truth_summary = db.SummaryTuple(
        n_logs=0, n_flight_segments=0, total_flight_time=dt.timedelta(), flight_segments=[]
    )

    assert db.summary_query() == truth_summary


SUMMARY_QUERIES = (
    (
        [DUMMY_FLIGHT_LOG],
        db.SummaryTuple(
            n_logs=1,
            n_flight_segments=1,
            total_flight_time=DUMMY_DURATION,
            flight_segments=[DUMMY_DURATION],
        ),
    ),
    (
        DUMMY_BULK_LOGS,
        db.SummaryTuple(
            n_logs=2,
            n_flight_segments=2,
            total_flight_time=dt.timedelta(seconds=10),
            flight_segments=[DUMMY_DURATION, DUMMY_DURATION],
        ),
    ),
    (
        [DUMMY_FLIGHT_LOG, DUMMY_FLIGHT_LOG_NO_FLIGHTS],
        db.SummaryTuple(
            n_logs=2,
            n_flight_segments=1,
            total_flight_time=DUMMY_DURATION,
            flight_segments=[DUMMY_DURATION],
        ),
    ),
)


@pytest.mark.parametrize(("flight_logs, truth_summary"), SUMMARY_QUERIES)
def test_summary_query(
    flight_logs: list[FlightLog], truth_summary: db.SummaryTuple, session: None
) -> None:
    db.bulk_insert(flight_logs)

    assert db.summary_query() == truth_summary


ppg_log/metrics.py
from __future__ import annotations

import datetime as dt
import typing as t
from collections import deque
from dataclasses import dataclass
from enum import IntEnum
from functools import cached_property, partial
from itertools import zip_longest
from pathlib import Path

import humanize
import numpy as np
import pandas as pd

from ppg_log import db, parser, viz
from ppg_log.exceptions import FlightSegmentationError

START_TRIM = 45  # seconds
ROLLING_WINDOW_WIDTH = 5
AIRBORNE_THRESHOLD = 2.235  # Groundspeed, m/s
FLIGHT_LENGTH_THRESHOLD = 15  # seconds

NUMERIC_T = int | float

LOG_DATETIME_FMT = r"%y-%m-%d_%H-%M-%S"

HUMANIZED_DELTA = partial(humanize.precisedelta, minimum_unit="seconds", format="%d")


class FlightMode(IntEnum):  # noqa: D101
    GROUND = 0
    AIRBORNE = 1


@dataclass(frozen=True, slots=True)
class FlightSegment:  # noqa: D101
    start_idx: int
    end_idx: int
    duration: dt.timedelta

    def __str__(self) -> str:  # pragma: no cover
        return (
            f"Start idx: {self.start_idx}\n"
            f"End idx: {self.end_idx}\n"
            f"Duration: {self.duration.seconds} seconds"
        )


@dataclass(slots=True)
class LogMetadata:  # noqa: D101
    log_date: str  # expected as YY-MM-DD
    log_time: str

    # Flight quantities are calculated downstream
    n_flight_segments: int | None = None  # If None, no metrics calculations have been done
    total_flight_time: dt.timedelta = dt.timedelta()
    flight_segments: list[FlightSegment] | None = None

    def __str__(self) -> str:  # pragma: no cover
        if self.flight_segments:
            humanized_time = HUMANIZED_DELTA(self.total_flight_time)
        else:
            humanized_time = "No flights detected"

        return (
            f"Log Date: {self.log_date} {self.log_time}\n"
            f"Flight Segments: {self.n_flight_segments}\n"
            f"Total Flight Time: {humanized_time}"
        )


@dataclass  # Can't slot w/cached_property
class FlightLog:  # noqa: D101
    flight_data: pd.DataFrame
    metadata: LogMetadata

    def __str__(self) -> str:  # pragma: no cover
        return str(self.metadata)

    @cached_property
    def log_datetime(self) -> dt.datetime:
        """Generate a `datetime` instance from the `FlightLog`'s metadata."""
        datestr = f"{self.metadata.log_date}_{self.metadata.log_time}"
        return dt.datetime.strptime(datestr, LOG_DATETIME_FMT)

    def summary_plot(
        self, show_plot: bool = True, save_dir: Path | None = None
    ) -> None:  # pragma: no cover
        """Build a summary plot for optional display and/or output to an image file."""
        if save_dir is not None:
            filename = f"{self.metadata.log_date}_{self.metadata.log_time}.png"
            save_path = save_dir / filename
        else:
            save_path = None

        viz.summary_plot(self, save_path=save_path, show_plot=show_plot)


@dataclass(frozen=True, slots=True)
class LogSummary:  # noqa: D101
    n_logs: int

    # If these are None, no metrics calculations have been done
    n_flight_segments: int | None
    total_flight_time: dt.timedelta | None
    avg_flight_time: dt.timedelta | None
    shortest_flight: dt.timedelta | None
    longest_flight: dt.timedelta | None

    def __str__(self) -> str:  # pragma: no cover
        if self.total_flight_time:
            total_time = HUMANIZED_DELTA(self.total_flight_time)
            avg_time = HUMANIZED_DELTA(self.avg_flight_time)
            shortest_time = HUMANIZED_DELTA(self.shortest_flight)
            longest_time = HUMANIZED_DELTA(self.longest_flight)

            humanized_time = (
                f"    Total Flight Time: {total_time}\n"
                f"    Average Flight Time: {avg_time}\n"
                f"    Shortest Flight: {shortest_time}\n"
                f"    Longest Flight: {longest_time}\n"
            )
        else:
            humanized_time = "    No flights detected or flight metrics not yet calculated."

        return f"Log Summary:\n    Flight Logs: {self.n_logs}\n{humanized_time}"

    @classmethod
    def from_flight_logs(cls, flight_logs: t.Collection[FlightLog]) -> LogSummary:
        """Generate a `LogSummary` instance from the provided `FlightLog` instances."""
        if not flight_logs:
            raise ValueError("No flight logs provided.")

        n_segments = 0
        flight_time = dt.timedelta()
        shortest = dt.timedelta()
        longest = dt.timedelta()
        for log in flight_logs:
            if log.metadata.n_flight_segments is not None:
                n_segments += log.metadata.n_flight_segments

                if log.metadata.flight_segments:  # pragma: no branch
                    for segment in log.metadata.flight_segments:
                        flight_time += segment.duration

                        if shortest == dt.timedelta() or segment.duration < shortest:
                            shortest = segment.duration

                        if segment.duration > longest:
                            longest = segment.duration
        else:
            if flight_time == dt.timedelta():
                total_flight_time = None
            else:
                total_flight_time = flight_time

        if total_flight_time is not None and n_segments is not None:
            avg_flight_time = dt.timedelta(seconds=total_flight_time.total_seconds() / n_segments)
            shortest_flight = shortest
            longest_flight = longest
        else:
            avg_flight_time = None
            shortest_flight = None
            longest_flight = None

        return cls(
            n_logs=len(flight_logs),
            n_flight_segments=n_segments if n_segments else None,
            total_flight_time=total_flight_time,
            avg_flight_time=avg_flight_time,
            shortest_flight=shortest_flight,
            longest_flight=longest_flight,
        )

    @classmethod
    def from_flight_log(cls, flight_log: FlightLog) -> LogSummary:
        """Generate a `LogSummary` instance from the provided `FlightLog` instance."""
        return cls.from_flight_logs([flight_log])

    @classmethod
    def from_db_query(cls, db_data: db.SummaryTuple) -> LogSummary:
        """Generate a `LogSummary` instance from the provided DB query response."""
        avg_flight_time = dt.timedelta(
            seconds=db_data.total_flight_time.total_seconds() / db_data.n_flight_segments
        )

        shortest = dt.timedelta()
        longest = dt.timedelta()
        for segment in db_data.flight_segments:
            if shortest == dt.timedelta() or segment < shortest:
                shortest = segment

            if segment > longest:
                longest = segment

        return cls(
            n_logs=db_data.n_logs,
            n_flight_segments=db_data.n_flight_segments,
            total_flight_time=db_data.total_flight_time,
            avg_flight_time=avg_flight_time,
            shortest_flight=shortest,
            longest_flight=longest,
        )


def _classify_flight_mode(groundspeed: float, airborne_threshold: NUMERIC_T) -> FlightMode:
    """Classify inflight vs. on ground based on the provided groundspeed threshold."""
    if groundspeed >= airborne_threshold:
        return FlightMode.AIRBORNE
    else:
        return FlightMode.GROUND


def classify_flight(
    flight_log: pd.DataFrame,
    window_width: int = ROLLING_WINDOW_WIDTH,
    airborne_threshold: NUMERIC_T = AIRBORNE_THRESHOLD,
) -> pd.DataFrame:
    """
    Classify inflight vs. on ground for the provided flight log based on groundspeed.

    To address noise in the groundspeed measurements, a rolling window mean of `window_width`
    groundspeeds is passed to the flight mode classifier.
    """
    flight_log["flight_mode"] = (
        flight_log["groundspeed"]
        .rolling(window_width, min_periods=1)
        .mean()
        .apply(_classify_flight_mode, airborne_threshold=airborne_threshold)
    )

    return flight_log


def _segment_flights(
    flight_data: pd.DataFrame, start_trim: NUMERIC_T, midair_start: bool
) -> tuple[list[list[int]], list[NUMERIC_T]]:
    """
    Identify candidate flight segments from the provided flight data.

    A list of `[takeoff, landing]` indices is returned along with a list of the time deltas, as
    decimal seconds, from the end of a segment to the beginning of the next segment.

    If `midair_start` is `True`, it's assumed that the first segment begins at the beginning of the
    file. In this case, `start_trim` will be set to zero upstream.
    """
    elapsed_time = flight_data["elapsed_time"]

    # Find the trim index
    trim_idx = (elapsed_time >= start_trim).idxmax()

    # Find consecutive runs of inflight modes & group by start & end indices of each run
    # AKA find takeoffs & landings
    diffs = np.abs(np.diff(flight_data["flight_mode"].iloc[trim_idx:]))
    diffs[0] = 0
    # Offset by the trim index since numpy's indices will be relative to the slice
    flights = np.flatnonzero(diffs == 1) + trim_idx

    # Add the start index if we're assuming the flight log starts in midair. Unless something else
    # is amiss, this should give us an even number of indices
    if midair_start:
        flights = np.concatenate(([0], flights))

    # Calculate time delta between flight segments, then reshape into nx2 for segment indices
    next_segment_delta = []
    for current_end, next_start in zip(
        elapsed_time.iloc[flights[1:-1:2]], elapsed_time.iloc[flights[2::2]]
    ):
        next_segment_delta.append(next_start - current_end)

    try:
        flights = flights.reshape(-1, 2)
    except ValueError:
        raise FlightSegmentationError(
            "Could not identify start and end indices for all flight segments."
        )

    # Cast flights to a list on return so I don't have to deal the huge cascade of mypy errors from
    # trying to use an ndarray with zip_longest
    return flights.tolist(), next_segment_delta


def find_flights(
    flight_data: pd.DataFrame,
    time_threshold: NUMERIC_T,
    start_trim: NUMERIC_T,
    midair_start: bool,
) -> list[FlightSegment] | None:
    """
    Identify start & end indices of flight segments for the provided flight log.

    Some basic filtering is done on candidate flight segments to help mitigate false positives from
    groundspeed instability.

    There are 2 classifications for short-duration segments:
        1. Transient spike while firmly on the ground, these should be discarded
        2. Noise while taking off, in flight, or while landing, these should be retained

    `time_threshold`, in seconds, is used to help classify short-duration segments as well as
    identify when the pilot has landed.

    `start_trim`, in seconds, is used to exclude segments from the beginning of the data file.

    If `midair_start` is `True`, it's assumed that the log file begins while airborne.
    """
    elapsed_time = flight_data["elapsed_time"]

    flight_segments, next_segment_delta = _segment_flights(
        flight_data=flight_data,
        start_trim=start_trim,
        midair_start=midair_start,
    )
    if len(flight_segments) == 0:
        return None

    valid_flights = []
    flight_indices: deque[int] = deque()
    for (segment_start, segment_end), next_delta in zip_longest(
        flight_segments, next_segment_delta
    ):
        segment_duration = elapsed_time.iloc[segment_end] - elapsed_time.iloc[segment_start]
        if next_delta is not None and (segment_duration < time_threshold):
            # Check for transient spikes
            # These are below the duration threshold & distant from the next flight segment
            if not flight_indices and (next_delta >= time_threshold):
                flight_indices.clear()
                continue
            else:
                # Otherwise, we'll consider this segment as part of the current flight segment
                flight_indices.extend((segment_start, segment_end))
        else:
            # This is a valid flight segment and/or the last segment in the file
            flight_indices.extend((segment_start, segment_end))

        # Now check the time to the next flight segment to see if we've landed
        if next_delta is None or (next_delta >= time_threshold):
            takeoff_idx = flight_indices[0]
            landing_idx = flight_indices[-1]
            flight_indices.clear()

            # Transient spikes may be close enough to combine into a segment that's shorter than the
            # time threshold, or may end up at the very end of the file
            # This can be discarded
            flight_duration = dt.timedelta(
                seconds=elapsed_time.iloc[landing_idx] - elapsed_time.iloc[takeoff_idx]
            )
            if flight_duration.total_seconds() < time_threshold:
                continue

            valid_flights.append(FlightSegment(takeoff_idx, landing_idx, flight_duration))

    if len(valid_flights) == 0:
        return None
    else:
        return valid_flights


def generate_flight_metrics(
    flight_log: FlightLog,
    airborne_threshold: NUMERIC_T = AIRBORNE_THRESHOLD,
    time_threshold: NUMERIC_T = FLIGHT_LENGTH_THRESHOLD,
    start_trim: NUMERIC_T = START_TRIM,
    midair_start: bool = False,
    classify_segments: bool = True,
) -> FlightLog:
    """Generate flight segment information for the provided `FlightLog` instance."""
    flight_log.flight_data = classify_flight(
        flight_log.flight_data, airborne_threshold=airborne_threshold
    )

    if classify_segments:
        flight_log.metadata.flight_segments = find_flights(
            flight_log.flight_data,
            time_threshold=time_threshold,
            start_trim=start_trim,
            midair_start=midair_start,
        )
        if flight_log.metadata.flight_segments:
            flight_log.metadata.n_flight_segments = len(flight_log.metadata.flight_segments)
            flight_log.metadata.total_flight_time = sum(
                (segment.duration for segment in flight_log.metadata.flight_segments),
                start=dt.timedelta(),
            )
        else:
            flight_log.metadata.n_flight_segments = 0
            flight_log.metadata.total_flight_time = dt.timedelta()

    return flight_log


def process_log(
    log_file: Path,
    start_trim: NUMERIC_T = START_TRIM,
    airborne_threshold: NUMERIC_T = AIRBORNE_THRESHOLD,
    time_threshold: NUMERIC_T = FLIGHT_LENGTH_THRESHOLD,
    midair_start: bool = False,
    classify_segments: bool = True,
) -> FlightLog:
    """
    Processing pipeline for an individual FlySight or Gaggle log file.

    NOTE: Processing pipeline selection between Flysight vs. Gaggle is done via case-insensitive
    extension matching: `*.CSV` for Flysight and `*.GPX` for Gaggle.
    """
    # Log files are grouped by date, need to retain this since it's not in the CSV filename
    log_date = log_file.parent.stem

    match log_file.suffix.upper():
        case ".CSV":
            log_time = log_file.stem
            flight_data = parser.load_flysight(log_file)
        case ".GPX":
            flight_data, log_datetime = parser.load_gaggle(log_file)
            log_time = log_datetime.strftime(r"%H-%M-%S")
        case _:
            raise ValueError(f"Unsupported file type: '{log_file.suffix}'")

    flight_log = FlightLog(
        flight_data=flight_data,
        metadata=LogMetadata(log_date=log_date, log_time=log_time),
    )

    # Force `start_trim` to zero if we're assuming the log starts while we're already airborne
    if midair_start:
        start_trim = 0

    flight_log = generate_flight_metrics(
        flight_log,
        airborne_threshold=airborne_threshold,
        time_threshold=time_threshold,
        start_trim=start_trim,
        midair_start=midair_start,
        classify_segments=classify_segments,
    )

    return flight_log


def batch_process(
    top_dir: Path,
    log_pattern: str = r"*.CSV",
    start_trim: NUMERIC_T = START_TRIM,
    airborne_threshold: NUMERIC_T = AIRBORNE_THRESHOLD,
    time_threshold: NUMERIC_T = FLIGHT_LENGTH_THRESHOLD,
    classify_segments: bool = True,
    verbose: bool = True,
) -> list[FlightLog]:
    """
    Batch process pipeline for a directory of FlySight or Gaggle logs.

    Log file discovery is not recursive by default, the `log_pattern` kwarg can be adjusted to
    support a recursive glob.

    NOTE: File case sensitivity is deferred to the OS; `log_pattern` is passed to glob as-is so
    matches may or may not be case-sensitive.
    """
    # Listify flight logs to get a total count
    log_files = list(top_dir.glob(log_pattern))
    if verbose:
        print(f"Found {len(log_files)} log files to process.")

    parsed_logs = []
    for log_file in log_files:
        if verbose:
            print(f"Processing {log_file.parent.stem}/{log_file.name} ... ", end="")

        try:
            flight_log = process_log(
                log_file,
                start_trim=start_trim,
                airborne_threshold=airborne_threshold,
                time_threshold=time_threshold,
                classify_segments=classify_segments,
            )
            parsed_logs.append(flight_log)
        except FlightSegmentationError:
            if verbose:
                print("Could not segment, skipped.")
            continue

        if verbose:
            print("Done")

    return parsed_logs
